testbeam_analysis/dut_alignment.py

```diff
@@ -858,35 +842,41 @@ def alignment(input_track_candidates_file, input_alignment_file, n_pixels, pixel
         alignment_parameters = _create_alignment_array(n_duts)
         alignment_parameters['translation_z'] = prealignment['z']
 
-        if initial_rotation:
-            if isinstance(initial_rotation[0], Iterable):
-                for dut_index in range(n_duts):
+        beta = np.arctan(prealignment[align_telescope[-1]]["column_c0"] / prealignment[align_telescope[-1]]["z"])
+        alpha = -np.arctan(prealignment[align_telescope[-1]]["row_c0"] / prealignment[align_telescope[-1]]["z"])
+        global_rotation = geometry_utils.rotation_matrix(alpha=alpha,
+                                                         beta=beta,
+                                                         gamma=0.0)
+
+        for dut_index in range(n_duts):
+            if (isinstance(initial_translation, Iterable) and initial_translation[dut_index] is None) or initial_translation is None:
+                if dut_index in align_telescope:
+                    # Telescope is aligned to these DUTs, so align them to the z-axis
+                    alignment_parameters['translation_x'][dut_index] = 0.0
+                    alignment_parameters['translation_y'][dut_index] = 0.0
+                else:
+                    # Correct x and y, so that the telescope is aligned to the z-axis
+                    aligned_offset = np.dot(global_rotation, np.array([prealignment[dut_index]["column_c0"], prealignment[dut_index]["row_c0"], prealignment[dut_index]["z"]]))
+                    alignment_parameters['translation_x'][dut_index] = aligned_offset[0]
+                    alignment_parameters['translation_y'][dut_index] = aligned_offset[1]
+            elif isinstance(initial_translation, Iterable) and isinstance(initial_translation[dut_index], Iterable) and len(initial_translation[dut_index]) in [2, 3]:
```

Comment:
You introduced a lot of data type error checks by hand. For this feature there are python packages like the rather famous [voluptuous](https://julien.danjou.info/blog/2015/python-schema-validation-voluptuous).